packages/algorand-js/src/interpreter/opcode-list.ts

```diff
@@ -1217,6 +1219,302 @@ export class Return extends Op {
+// params: account index, application id (top of the stack on opcode entry).
+// Return: 1 if opted in and 0 otherwise.
+export class AppOptedIn extends Op {
+  readonly interpreter: Interpreter;
+
+  /**
+   * @param args words list extracted from line
+   * @param line line number in TEAL file
+   * @param interpreter interpreter object
+   */
+  constructor (args: string[], line: number, interpreter: Interpreter) {
+    super();
+    assertLen(args.length, 0, line);
+    this.interpreter = interpreter;
+  }
+
+  execute (stack: TEALStack): void {
+    this.assertMinStackLen(stack, 2);
+    const appId = this.assertBigInt(stack.pop());
+    const accountIndex = this.assertBigInt(stack.pop());
+
+    const account = this.getAccount(accountIndex, this.interpreter);
+    const localState = account["apps-local-state"];
+
+    const isOptedIn = localState.find(state => state.id === Number(appId));
+    if (isOptedIn) {
+      stack.push(BIGINT1);
+    } else {
+      stack.push(BIGINT0);
+    }
+  }
+}
+
+// read from account specified by Txn.Accounts[A] from local state of the current application key B => value
+export class AppLocalGet extends Op {
+  readonly interpreter: Interpreter;
+
+  /**
+   * @param args words list extracted from line
+   * @param line line number in TEAL file
+   * @param interpreter interpreter object
+   */
+  constructor (args: string[], line: number, interpreter: Interpreter) {
+    super();
+    assertLen(args.length, 0, line);
+    this.interpreter = interpreter;
+  }
+
+  execute (stack: TEALStack): void {
+    this.assertMinStackLen(stack, 2);
+    const key = this.assertBytes(stack.pop());
+    const accountIndex = this.assertBigInt(stack.pop());
+
+    const account = this.getAccount(accountIndex, this.interpreter);
+    const appId = this.interpreter.tx.apid;
+
+    const val = getLocalState(appId, account, key);
+    if (val) {
+      stack.push(val);
+    } else {
+      stack.push(BIGINT0); // The value is zero if the key does not exist.
+    }
+  }
+}
+
+// read from account specified by Txn.Accounts[A] from local state of the
+// application B key C => {0 or 1 (top), value}
+export class AppLocalGetEx extends Op {
+  readonly interpreter: Interpreter;
+
+  /**
+   * @param args words list extracted from line
+   * @param line line number in TEAL file
+   * @param interpreter interpreter object
+   */
+  constructor (args: string[], line: number, interpreter: Interpreter) {
+    super();
+    assertLen(args.length, 0, line);
+    this.interpreter = interpreter;
+  }
+
+  execute (stack: TEALStack): void {
+    this.assertMinStackLen(stack, 3);
+    const key = this.assertBytes(stack.pop());
+    const appId = this.assertBigInt(stack.pop());
+    const accountIndex = this.assertBigInt(stack.pop());
+
+    const account = this.getAccount(accountIndex, this.interpreter);
+    const val = getLocalState(Number(appId), account, key);
+    if (val) {
+      stack.push(val);
```

Comment:
Please push did_exist flag in stack. (according to docstring)

Reply:
yes, sorry